### PR TITLE
Feature/replace with peer into command ir 789

### PR DESCRIPTION
**Pull request by Alexey-N-Chernyshov**
### Description of the Change
Replace address and pubkey with peer in addPeer command. Add Peer object to shared_model.

### Benefits
Make shared_model more complete.

### Possible Drawbacks 
Thorough review is required.

### Usage Examples or Tests *[optional]*
Tests are included into commit.

### CODE REVIEW

File: libs/common/types.hpp [hunk 1]
```diff
@@ -303,5 +304,8 @@ namespace iroha {
     return typeid(Base) == typeid(ptr);
   }
 
+  // Peer address
+  using AddressType = std::string;
```
Comment (lebdron) [inline libs/common/types.hpp]: It does not seem necessary, since old model can use shared model.

File: libs/common/types.hpp [hunk 2]
```diff
@@ -184,7 +184,8 @@ namespace iroha {
    * @return monadic value, which can be of another type
    */
   template <typename T, typename Transform>
-  auto operator|(T t, Transform f) -> typename std::enable_if<
+  auto operator|(T t, Transform f) -> typename std::
```
Comment (lebdron) [inline libs/common/types.hpp]: clang-format seems broken here.

File: shared_model/backend/protobuf/common_objects/peer.hpp [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+      using Lazy = detail::LazyInitializer<T>;
+
+      const Lazy<interface::types::PubkeyType> public_key_;
+
+      const Lazy<BlobType> blob_;
+    };
+
+  }  // namespace proto
+}  // namespace shared_model
+#endif  // IROHA_PROTO_PEER_HPP
```
Comment (lebdron) [inline shared_model/backend/protobuf/common_objects/peer.hpp]: IROHA_PROTO_PEER_HPP -> IROHA_SHARED_MODEL_PROTO_PEER_HPP

File: shared_model/interfaces/common_objects/peer.hpp [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+        OldModelType *oldModel = new OldModelType();
+        oldModel->address = address();
+        oldModel->pubkey = pubkey().makeOldModel<decltype(oldModel->pubkey)>();
+        return oldModel;
+      }
+    };
+  }
+}
+
+#endif  // IROHA_PEER_HPP
```
Comment (lebdron) [inline shared_model/interfaces/common_objects/peer.hpp]: IROHA_PEER_HPP -> IROHA_SHARED_MODEL_PEER_HPP

File: shared_model/interfaces/common_objects/peer.hpp [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+#include "common/types.hpp"
+#include "interfaces/base/hashable.hpp"
+#include "interfaces/common_objects/types.hpp"
+#include "model/peer.hpp"
+
+namespace shared_model {
+  namespace interface {
+
+    /**
+     * Peer is Model, which contains information about network participants
```
Comment (lebdron) [inline shared_model/interfaces/common_objects/peer.hpp]: Maybe "Representation of network participants" should suffice?

File: irohad/model/peer.hpp [hunk 6]
```diff
@@ -47,7 +45,12 @@ namespace iroha {
         } else {
           return false;
         }
-      };
+      }
+
+      Peer() = default;
+
+      Peer(const AddressType &address, const pubkey_t &pubkey)
```
[the remark this answers is not in the document]
Comment (Alexey-N-Chernyshov) [reply irohad/model/peer.hpp]: What is KeyType? Old model uses `pubkey_t`.

File: libs/common/types.hpp (hunk 1 above)
Comment (Alexey-N-Chernyshov) [reply libs/common/types.hpp]: mb it worth to keep it both in the old model and in the shared model?

File: test/module/shared_model/validators/field_validator_test.cpp [hunk 7]
```diff
@@ -244,38 +257,64 @@ class FieldValidatorTest : public ValidatorsTest {
...
   }
 
-  // Address validation test is handled in libs/validator,
-  // so test cases are not exhaustive
-  std::vector<FieldTestCase> address_test_cases{
-      makeValidCase(&FieldValidatorTest::address_localhost, "182.13.35.1:3040"),
-      invalidAddressTestCase("invalid_ip_address", "182.13.35.1:3040--"),
-      invalidAddressTestCase("empty_string", "")};
+  /// Make test case for invalid peer public key
+  FieldTestCase makeInvalidPPeerPubkeyTestCase(const std::string &case_name,
```
Comment (lebdron) [inline test/module/shared_model/validators/field_validator_test.cpp]: makeInvalidPPeerPubkeyTestCase -> makeInvalidPeerPubkeyTestCase

File: shared_model/interfaces/common_objects/peer.hpp [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+#include "interfaces/common_objects/types.hpp"
+#include "model/peer.hpp"
+
+namespace shared_model {
+  namespace interface {
+
+    /**
+     * Representation of a network participant.
+     */
+    class Peer : public Hashable<Peer, iroha::model::Peer> {
```
Comment (lebdron) [inline shared_model/interfaces/common_objects/peer.hpp]: Changes regarding backward compatibility from develop should be applied here.

**Comment by bakhtin** [comment]
SonarQube analysis reported 188 issues
* ![CRITICAL][CRITICAL] 1 critical
* ![MINOR][MINOR] 60 minor
* ![INFO][INFO] 127 info

#### Top 10 issues

1. ![CRITICAL][CRITICAL] [pb_command_factory.cpp#L578](https://github.com/hyperledger/iroha/blob/0f7f8b64d915f44b0117d8d531453e8f5a298af7/irohad/model/converters/impl/pb_command_factory.cpp#L578): The Cyclomatic Complexity of this function is 17 which is greater than 8 authorized. [rule](https://sonar.soramitsu.co.jp/coding_rules#rule_key=cxx%3AFunctionComplexity)
1. ![MINOR][MINOR] [interactive_transaction_cli.cpp#L196](https://github.com/hyperledger/iroha/blob/0f7f8b64d915f44b0117d8d531453e8f5a298af7/iroha-cli/interactive/impl/interactive_transaction_cli.cpp#L196): Define a constant instead of duplicating this literal false 3 times. [rule](https://sonar.soramitsu.co.jp/coding_rules#rule_key=cxx%3AStringLiteralDuplicated)
1. ![MINOR][MINOR] [interactive_transaction_cli.cpp#L312](https://github.com/hyperledger/iroha/blob/0f7f8b64d915f44b0117d8d531453e8f5a298af7/iroha-cli/interactive/impl/interactive_transaction_cli.cpp#L312): The function 'parseRevokePermission' is never used. [rule](https://sonar.soramitsu.co.jp/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [json_command_factory.cpp#L49](https://github.com/hyperledger/iroha/blob/0f7f8b64d915f44b0117d8d531453e8f5a298af7/irohad/model/converters/impl/json_command_factory.cpp#L49): Define a constant instead of duplicating this literal "value" 6 times. [rule](https://sonar.soramitsu.co.jp/coding_rules#rule_key=cxx%3AStringLiteralDuplicated)
1. ![MINOR][MINOR] [json_command_factory.cpp#L64](https://github.com/hyperledger/iroha/blob/0f7f8b64d915f44b0117d8d531453e8f5a298af7/irohad/model/converters/impl/json_command_factory.cpp#L64): Define a constant instead of duplicating this literal "precision" 6 times. [rule](https://sonar.soramitsu.co.jp/coding_rules#rule_key=cxx%3AStringLiteralDuplicated)
1. ![MINOR][MINOR] [json_command_factory.cpp#L76](https://github.com/hyperledger/iroha/blob/0f7f8b64d915f44b0117d8d531453e8f5a298af7/irohad/model/converters/impl/json_command_factory.cpp#L76): Define a constant instead of duplicating this literal "address" 2 times. [rule](https://sonar.soramitsu.co.jp/coding_rules#rule_key=cxx%3AStringLiteralDuplicated)
1. ![MINOR][MINOR] [json_command_factory.cpp#L77](https://github.com/hyperledger/iroha/blob/0f7f8b64d915f44b0117d8d531453e8f5a298af7/irohad/model/converters/impl/json_command_factory.cpp#L77): Define a constant instead of duplicating this literal "peer_key" 2 times. [rule](https://sonar.soramitsu.co.jp/coding_rules#rule_key=cxx%3AStringLiteralDuplicated)
1. ![MINOR][MINOR] [json_command_factory.cpp#L91](https://github.com/hyperledger/iroha/blob/0f7f8b64d915f44b0117d8d531453e8f5a298af7/irohad/model/converters/impl/json_command_factory.cpp#L91): Variable 'serializers_' is assigned in constructor body. Consider performing initialization in initialization list. [rule](https://sonar.soramitsu.co.jp/coding_rules#rule_key=cppcheck%3AuseInitializationList)
1. ![MINOR][MINOR] [json_command_factory.cpp#L118](https://github.com/hyperledger/iroha/blob/0f7f8b64d915f44b0117d8d531453e8f5a298af7/irohad/model/converters/impl/json_command_factory.cpp#L118): Define a constant instead of duplicating this literal "AddAssetQuantity" 2 times. [rule](https://sonar.soramitsu.co.jp/coding_rules#rule_key=cxx%3AStringLiteralDuplicated)
1. ![MINOR][MINOR] [json_command_factory.cpp#L120](https://github.com/hyperledger/iroha/blob/0f7f8b64d915f44b0117d8d531453e8f5a298af7/irohad/model/converters/impl/json_command_factory.cpp#L120): Define a constant instead of duplicating this literal "SubtractAssetQuantity" 2 times. [rule](https://sonar.soramitsu.co.jp/coding_rules#rule_key=cxx%3AStringLiteralDuplicated)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: test/module/shared_model/validators/field_validator_test.cpp [hunk 9]
```diff
@@ -174,6 +174,21 @@ class FieldValidatorTest : public ValidatorsTest {
 
   /************************** TEST CASES ***************************/
 
+  /// Make message for wrong formed field
```
Comment (muratovv) [inline test/module/shared_model/validators/field_validator_test.cpp]: Better to write extended documentation for the function

File: test/module/shared_model/validators/field_validator_test.cpp [hunk 10]
```diff
@@ -174,6 +174,21 @@ class FieldValidatorTest : public ValidatorsTest {
...
+  std::string makeMessageWrongField(const std::string &field_name,
+                                    const std::string &value) {
+    return (boost::format("Wrongly formed %s, passed value: '%s'") % field_name
+            % value)
+        .str();
+  }
+
+  /// Make message for wrong key size
```
Comment (muratovv) [inline test/module/shared_model/validators/field_validator_test.cpp]: Same as before

File: test/module/shared_model/validators/field_validator_test.cpp [hunk 11]
```diff
@@ -244,38 +257,64 @@ class FieldValidatorTest : public ValidatorsTest {
        false,
        "Amount must be greater than 0, passed value: 0"}};
 
-  FieldTestCase invalidAddressTestCase(const std::string &case_name,
-                                       const std::string &address) {
-    return makeInvalidCase(case_name,
-                           "peer address",
-                           &FieldValidatorTest::address_localhost,
-                           address);
+  /// Make test case for invalid peer address
```
Comment (muratovv) [inline test/module/shared_model/validators/field_validator_test.cpp]: Same as before

File: test/module/shared_model/validators/field_validator_test.cpp [hunk 12]
```diff
@@ -244,38 +257,64 @@ class FieldValidatorTest : public ValidatorsTest {
...
+            makeMessageWrongField("peer address", address)};
   }
 
-  // Address validation test is handled in libs/validator,
-  // so test cases are not exhaustive
-  std::vector<FieldTestCase> address_test_cases{
-      makeValidCase(&FieldValidatorTest::address_localhost, "182.13.35.1:3040"),
-      invalidAddressTestCase("invalid_ip_address", "182.13.35.1:3040--"),
-      invalidAddressTestCase("empty_string", "")};
+  /// Make test case for invalid peer public key
```
Comment (muratovv) [inline test/module/shared_model/validators/field_validator_test.cpp]: s.a.b.

**Comment by bakhtin** [comment]
SonarQube analysis reported 188 issues
* ![CRITICAL][CRITICAL] 1 critical
* ![MINOR][MINOR] 60 minor
* ![INFO][INFO] 127 info

#### Top 10 issues

1. ![CRITICAL][CRITICAL] [pb_command_factory.cpp#L578](https://github.com/hyperledger/iroha/blob/a476dc71fb2374ab5206a51f7a655a567bedc7ea/irohad/model/converters/impl/pb_command_factory.cpp#L578): The Cyclomatic Complexity of this function is 17 which is greater than 8 authorized. [rule](https://sonar.soramitsu.co.jp/coding_rules#rule_key=cxx%3AFunctionComplexity)
1. ![MINOR][MINOR] [interactive_transaction_cli.cpp#L196](https://github.com/hyperledger/iroha/blob/a476dc71fb2374ab5206a51f7a655a567bedc7ea/iroha-cli/interactive/impl/interactive_transaction_cli.cpp#L196): Define a constant instead of duplicating this literal false 3 times. [rule](https://sonar.soramitsu.co.jp/coding_rules#rule_key=cxx%3AStringLiteralDuplicated)
1. ![MINOR][MINOR] [interactive_transaction_cli.cpp#L312](https://github.com/hyperledger/iroha/blob/a476dc71fb2374ab5206a51f7a655a567bedc7ea/iroha-cli/interactive/impl/interactive_transaction_cli.cpp#L312): The function 'parseRevokePermission' is never used. [rule](https://sonar.soramitsu.co.jp/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [json_command_factory.cpp#L49](https://github.com/hyperledger/iroha/blob/a476dc71fb2374ab5206a51f7a655a567bedc7ea/irohad/model/converters/impl/json_command_factory.cpp#L49): Define a constant instead of duplicating this literal "value" 6 times. [rule](https://sonar.soramitsu.co.jp/coding_rules#rule_key=cxx%3AStringLiteralDuplicated)
1. ![MINOR][MINOR] [json_command_factory.cpp#L64](https://github.com/hyperledger/iroha/blob/a476dc71fb2374ab5206a51f7a655a567bedc7ea/irohad/model/converters/impl/json_command_factory.cpp#L64): Define a constant instead of duplicating this literal "precision" 6 times. [rule](https://sonar.soramitsu.co.jp/coding_rules#rule_key=cxx%3AStringLiteralDuplicated)
1. ![MINOR][MINOR] [json_command_factory.cpp#L76](https://github.com/hyperledger/iroha/blob/a476dc71fb2374ab5206a51f7a655a567bedc7ea/irohad/model/converters/impl/json_command_factory.cpp#L76): Define a constant instead of duplicating this literal "address" 2 times. [rule](https://sonar.soramitsu.co.jp/coding_rules#rule_key=cxx%3AStringLiteralDuplicated)
1. ![MINOR][MINOR] [json_command_factory.cpp#L77](https://github.com/hyperledger/iroha/blob/a476dc71fb2374ab5206a51f7a655a567bedc7ea/irohad/model/converters/impl/json_command_factory.cpp#L77): Define a constant instead of duplicating this literal "peer_key" 2 times. [rule](https://sonar.soramitsu.co.jp/coding_rules#rule_key=cxx%3AStringLiteralDuplicated)
1. ![MINOR][MINOR] [json_command_factory.cpp#L91](https://github.com/hyperledger/iroha/blob/a476dc71fb2374ab5206a51f7a655a567bedc7ea/irohad/model/converters/impl/json_command_factory.cpp#L91): Variable 'serializers_' is assigned in constructor body. Consider performing initialization in initialization list. [rule](https://sonar.soramitsu.co.jp/coding_rules#rule_key=cppcheck%3AuseInitializationList)
1. ![MINOR][MINOR] [json_command_factory.cpp#L118](https://github.com/hyperledger/iroha/blob/a476dc71fb2374ab5206a51f7a655a567bedc7ea/irohad/model/converters/impl/json_command_factory.cpp#L118): Define a constant instead of duplicating this literal "AddAssetQuantity" 2 times. [rule](https://sonar.soramitsu.co.jp/coding_rules#rule_key=cxx%3AStringLiteralDuplicated)
1. ![MINOR][MINOR] [json_command_factory.cpp#L120](https://github.com/hyperledger/iroha/blob/a476dc71fb2374ab5206a51f7a655a567bedc7ea/irohad/model/converters/impl/json_command_factory.cpp#L120): Define a constant instead of duplicating this literal "SubtractAssetQuantity" 2 times. [rule](https://sonar.soramitsu.co.jp/coding_rules#rule_key=cxx%3AStringLiteralDuplicated)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request merged**